File: MATTHEW/leetcode/AIcoding/4.GPT-4o/java_solutions/2967-minimum-cost-to-make-array-equalindromic.java
class Solution {
    public int minCost(int[] A) {
        int n = A.length;
        int[][] dp = new int[n][n];

        for (int len = 2; len <= n; len++) {
            for (int i = 0; i <= n - len; i++) {
                int j = i + len - 1;
                if (A[i] == A[j]) {
                    dp[i][j] = dp[i + 1][j - 1];
                } else {
                    dp[i][j] = Math.min(dp[i + 1][j] + A[i], dp[i][j - 1] + A[j]);
                }
            }
        }

        return dp[0][n - 1];
    }
}

File: MATTHEW/leetcode/AIcoding/5.Qwen-3/cpp_solutions/2967-minimum-cost-to-make-array-equalindromic.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

class Solution {
public:
    int minCostToMakeArrayEqualindromic(vector<int>& nums) {
        int n = nums.size();
        vector<int> sorted_nums = nums;
        sort(sorted_nums.begin(), sorted_nums.end());

        auto isPalindrome = [](int x) {
            int rev = 0, original = x;
            while (x > 0) {
                rev = rev * 10 + x % 10;
                x /= 10;
            }
            return rev == original;
        };

        auto getCost = [&](int target) {
            int cost = 0;
            for (int num : nums) {
                cost += abs(num - target);
            }
            return cost;
        };

        int min_cost = INT_MAX;
        for (int i = 0; i < n; ++i) {
            int mid = sorted_nums[i];
            for (int j = 0; j <= 1; ++j) {
                int candidate = mid + j;
                if (isPalindrome(candidate)) {
                    min_cost = min(min_cost, getCost(candidate));
                }
            }
        }

        return min_cost;
    }
};